hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java

```diff
@@ -947,17 +1048,529 @@ public void migrate(PartitionReplica source, int sourceCurrentReplicaIndex, int
+                return null;
+            }
+            if (migration.getSource() != null) {
+                PartitionReplica source = migration.getSource();
+                if (node.getClusterService().getMember(source.address(), source.uuid()) == null) {
+                    logger.fine("Source is not a member anymore. Ignoring " + migration);
+                    triggerRepartitioningAfterMigrationFailure();
+                    return null;
+                }
+            }
+            PartitionReplica destination = migration.getDestination();
+            if (node.getClusterService().getMember(destination.address(), destination.uuid()) == null) {
+                logger.fine("Destination is not a member anymore. Ignoring " + migration);
+                triggerRepartitioningAfterMigrationFailure();
+                return null;
+            }
+            return partitionOwner;
+        }
+
+        /** Returns the partition owner or {@code null} if it is not set. */
+        private Member getPartitionOwner() {
+            InternalPartitionImpl partition = partitionStateManager.getPartitionImpl(migration.getPartitionId());
+            PartitionReplica owner = partition.getOwnerReplicaOrNull();
+            if (owner == null) {
+                if (migration.isValid()) {
+                    logger.severe("Skipping migration! Partition owner is not set! -> partitionId="
+                            + migration.getPartitionId()
+                            + ", " + partition + " -VS- " + migration);
+                }
+                return null;
+            }
+            return node.getClusterService().getMember(owner.address(), owner.uuid());
+        }
+
+        /**
+         * Sends a {@link MigrationRequestOperation} to the {@code fromMember} and returns the migration result if the
+         * migration was successful.
+         */
+        @SuppressWarnings({"checkstyle:npathcomplexity", "checkstyle:cyclomaticcomplexity"})
+        private CompletionStage<Boolean> executeMigrateOperation(Member fromMember) {
+            long start = Timer.nanos();
+            CompletableFuture<Boolean> future;
+            try {
+                beforeMigration();
+
+                List<MigrationInfo> completedMigrations = getCompletedMigrations(migration.getPartitionId());
+                Operation op = new MigrationRequestOperation(migration, completedMigrations, 0, fragmentedMigrationEnabled);
+                future = nodeEngine.getOperationService()
+                        .createInvocationBuilder(SERVICE_NAME, op, fromMember.getAddress())
+                        .setCallTimeout(partitionMigrationTimeout)
+                        .invoke();
+            } catch (Throwable t) {
+                Level level = migration.isValid() ? Level.WARNING : Level.FINE;
+                logger.log(level, "Error during " + migration, t);
+                future = new CompletableFuture<>();
+                future.completeExceptionally(t);
```

Comment:
minor: could also use the static factory methods from `InternalCompletableFuture` for completed futures